### PR TITLE
Implement `ecli-api-server`

### LINKED ISSUE #157: [FEATURE] Add ecli api server for deploy eBPF programs
**Is your feature request related to a problem? Please describe.**

Add API for deploy eBPF programs from remote. For example, it could be used in k8s, android etc...

We have done this in eunomia-bpf v0.1.0, but it's removed now.

It should be:

1. using GPT to generate a swagger OpenAPI document
2. using swagger to generate client and server code for rust

It should be able to:

- upload/deploy an eBPF program to server
- load and attach an eBPF program to kernel
- delete the eBPF program from remote.

### PR DESCRIPTION
**Pull request by Officeyutong**
Closes #157 

Notice to the reviewer: Please ignore the `server-codegen` folder. Everything in that folder was generated by `openapi-generator`.

This PR rewrites most of the `ecli`'s code, and complete the `ecli api server`.

It implements an API server, which serves the api defined by openapi defs `apis.yaml`. The server implementation can be seen at `ecli-lib/src/runner/server_http`. It also provides a cli wrapper, in the `cli` folder, which can be directly executed and serve the API.

It also splits the original ecli project into several crates:
- `ecli-lib`: The core implementation, with several features specifying whether to enable the corresponding `client` or `server` codes
- `client`: The client CLI wrapper. It can act as both a native client (which runs programs natively, controlled by feature `native`) and http client(which runs program through the HTTP API on the remote machine, controlled by feature `http`)
- `server`: The server CLI wrapper. Can be directly executed and serve the API.

Note: I don't know how to let GitGuardian ignore the private keys in the `server-codegen` folder. The keys are randomly generated by the code generator, and has no realise usage.

TODO:
- [x] JSON runner
- [x] Tidy & clean the exist code
- [x] Implement server
- [x] More tests
- [x] Make CI happy
- [x] Docs

### CODE REVIEW

File: ecli/tests/api-test.py
Comment (ocfox) [inline ecli/tests/api-test.py:1]:
```suggestion
#!/usr/bin/env python
import os
```

File: ecli/client/src/http_client.rs
Comment (Officeyutong) [reply ecli/client/src/http_client.rs:60]: Port is included in the url

File: .github/workflows/ecli.yaml
Comment (Officeyutong) [reply .github/workflows/ecli.yaml:105]: To make the root user can execute cargo properly

File: ecli/ecli-lib/src/config/mod.rs
Comment (Officeyutong) [reply ecli/ecli-lib/src/config/mod.rs:34]: This is something the previous maintainer wrote. I'll fix them

File: ecli/client/src/main.rs
Comment (Officeyutong) [reply ecli/client/src/main.rs:4]: What to discuss

File: ecli/client/README.md
Comment (Officeyutong) [reply ecli/client/README.md:49]: Usually we use endpoint to call a HTTP API Service URL. Multiple ecli api server can be served at the same host(e.g when the server hosts a reversed-prroxy), so we need port to distinguish the server

Comment (Officeyutong) [reply ecli/client/README.md:49]: Usually we use endpoint to call a HTTP API Service URL. Multiple ecli api server can be served at the same host(e.g when the server hosts a reversed-prroxy), so we need port to distinguish the server

File: ecli/client/src/http_client.rs
Comment (Officeyutong) [reply ecli/client/src/http_client.rs:178]: - The new line is needed, since the logs previously printed may not be ended with a new line
- Originally I think format strings like "\nNo logs fetched" looks ugly and confusing. what do you think?

Comment (ocfox) [reply ecli/client/src/http_client.rs:178]: I think it should be a common practice, see
https://github.com/search?q=println%21%28%22%5Cn&type=code

File: ecli/Makefile
Comment (Officeyutong) [reply ecli/Makefile:8]: Do you mean to use a feature to control whether to build "remote-only client"?

File: ecli/client/src/http_client.rs
Comment (Officeyutong) [reply ecli/client/src/http_client.rs:178]: > I think it should be a common practice, see https://github.com/search?q=println%21%28%22%5Cn&type=code

Ok. I'll apply it

Comment (Officeyutong) [reply ecli/client/src/http_client.rs]: Fixed now

Comment (Officeyutong) [reply ecli/client/src/http_client.rs]: But there is another problem: Different platforms have dirrerent line breaks chars, `\r\n` on Windows and `\n` on *nix. Remote clients can be run under Windows. Should we take this plarform-specified behavior into consideration? on just use `\n` directly?

File: ecli/Makefile
Comment (ocfox) [reply ecli/Makefile:8]: yep, I think the default build should contain the power to run ebpf locally.

File: ecli/client/README.md
Comment (ocfox) [reply ecli/client/README.md:49]: ok, thanks for explaining

File: ecli/Makefile
Comment (Officeyutong) [reply ecli/Makefile]: Fixed now. Now `ecli` has been default to enable the `native` feature

File: ecli/client/src/http_client.rs
Comment (Officeyutong) [reply ecli/client/src/http_client.rs]: > But there is another problem: Different platforms have dirrerent line breaks chars, `\r\n` on Windows and `\n` on *nix. Remote clients can be run under Windows. Should we take this plarform-specified behavior into consideration? on just use `\n` directly?

Infact, use `println!()` is a good way to sole such platform differences

Comment (ocfox) [reply ecli/client/src/http_client.rs]: sorry, I don't know this diff on Windows. In that case, let's keep the original way.

Comment (Officeyutong) [reply ecli/client/src/http_client.rs:174]: Ok, I added that println!() back

**Review by yunwei37 — APPROVED** [review]

**Comment by yunwei37** [comment]
please fix the document links before merge.

**Comment by Officeyutong** [comment]
> please fix the document links before merge.

Seems that the documents issues are not related to this pull request

File: ecli/ecli-lib/src/helper.rs
Comment (ocfox) [inline ecli/ecli-lib/src/helper.rs:23]:
```suggestion
pub fn get_eunomia_home() -> Result<String> {
    let eunomia_home = std::env::var(EUNOMIA_HOME_ENV);
    if let Ok(home) = eunomia_home {
        Ok(home)
    } else if let Some(home) = home::home_dir() {
        let home = home.join(".eunomia");
        if !home.exists() {
            std::fs::create_dir_all(&home).map_err(Error::IOErr)?;
        }
        Ok(home.to_string_lossy().to_string())
    } else {
        Err(Error::Other(
            "home dir not found. Please set EUNOMIA_HOME env.".to_string(),
        ))
    }
}
```
It might look better (cleaner) this way. Personally, if there is a bad place, it is fine not to change it.

**Review by ocfox — COMMENTED** [review]
Almost, the rest I can not see what, except the codegen part of the code is very difficult to read, I did not review

**Comment by Officeyutong** [comment]
> Almost, the rest I can not see what, except the codegen part of the code is very difficult to read, I did not review

No need to review the generated code

**Pull request merged**
